### PR TITLE
Correctly implement QUIC sniffer when handling multiple initial packets

### PR DESCRIPTION
**Pull request by Vigilans**
In recent chromium browser, a QUIC connection may contain multiple initial packets. This PR fixes this issue, to:
1. Cache and parse multiple UDP packets until a ClientHello can be detected from CRYPTO stream.
2. Increase max sniffing packets for UDP from 2 to 5, to avoid missing any initial packet after 2 packets  (+3 simply because TCP handshake contains 3 packets, so the 2th sniffed packet for TCP is actually 5th packet)
3. `io.EOF` will not cause sniffing to fail. Though it maybe irrelavent for QUIC here because UDP packets have determined boundary. This fix may be helpful for TCP stream.

This fix has been proved to work without error for around 9 months in my local setup.

Addresses https://github.com/v2fly/v2ray-core/issues/3303

### CODE REVIEW

File: common/protocol/quic/sniff.go [hunk 1]
```diff
@@ -48,206 +48,226 @@ var (
...
-	if err != nil {
-		return nil, errNotQuic
-	}
+	// Parse QUIC packets
+	for len(b) > 0 {
+		buffer := buf.FromBytes(b)
+		typeByte, err := buffer.ReadByte()
+		if err != nil {
+			return nil, errNotQuic
+		}
```
Comment (Vigilans) [inline common/protocol/quic/sniff.go:62]: Most of changes in this file are just indented under `for len(b) > 0 {` block.

**Review by xiaokangwang — COMMENTED** [review]
I have reviewed the code and think it is in general nice, although I did find it failed to process one corner case when I was trying to add the test(pacp attached).

I am happy to merge it once the EOF thing is resolved.

Anther thing I think we don't need to worry too much but should keep an eye for is that because we are waiting for 100 ms for receiving the packet for sniffing, the worst case time out is now increased to 500ms. It is still too early to say it would break anything, but I am ready to fix this will a separate timeout timer as needed. 

[quic-chromebook-2.zip](https://github.com/user-attachments/files/18789598/quic-chromebook-2.zip)

File: common/protocol/quic/sniff_test.go [hunk 2]
```diff
@@ -126,7 +126,7 @@ func TestSniffQUICComplex(t *testing.T) {
...
...f24e69fdb883e850714fe27a45f43883486a986574fd1fc10f259fe90786441554514c8dade1f3b86fdaf5f54ab655e2d803c98aa56073b00c32148a1ed367dff3a2bd934ecba55141389990b661bbd9ce1ef1def13747d45500daf92cec9e60908274703e761cd46affd46622f2a2192a79425ebf51c875fc7ba3598e15e0ba2465fc3e87c8a5da1915d3b8abe4b16d21259f311183eee1e7d2b808a91a7c89b284df0eb6a2a79c610bbe47722b3e04d5a6c0e574816a94d97349b6976010eb8c7debf42210982f78de482b7cd068051bf57908dbf46b5ceaf64f5fb33ede4412c1ce81eb1dfb4e99e10dd9b57ebc6e62ecbf4ee2db04d9e48c62bd45f8fa51704d414296a2d51d25ced6a192034a44c67e09d8985b573f98e03fa36dd8dcce2c04b4d5b1f276b6a642aadbdcafcf09de1d234bf8bbbf64aeadf01519ddafb419b3e62d204e04c3d7ebaf54b09e387ac3e9c4781c11625a2f44fddb7a1886f21929bd01c283f64903b6ccbb463984dfadc00f6af2a421517da023fd319f528195cac5fe907624b70c0172479d07d78e266dbf20ab8fc302228f279ffdba7395a839c4a9d7a4e001a260e1702393968f1e9722f023b204cf09cfee9a7bba045e4a2a449ee9fbb5c36e93028cfc87a2e34914b1b4f01beeded175ac0fa73fea9292f2bc3b1247164d8e05cddc3981bdc24e5f596571c418f6fa00fd9d4d0898cbf0d2f5413bed5f100f1854903017b6bc88bd7e303b5e0e2417bbcc984731128eda550d31f9af0e6e743eb6916466bbd435617d56fa60b05cd7dca66a9f6f4be23d3c5ff5d900822c6d1d8d71b0bab24f57d9682381a87c",
```
Comment (Vigilans) [inline common/protocol/quic/sniff_test.go:129]: > I have reviewed the code and think it is in general nice, although I did find it failed to process one corner case when I was trying to add the test(pacp attached).

In this testcase (`QUIC Chromebook Handshake[2]; packet 1-3`), the third packet's hex included the UDP header (28 bytes) by mistake, making the payload length to be 1278 instead of 1250 (all three initial packet payload is 1250 bytes long). After fixing it the test passed.

**Comment by Vigilans** [comment]
> Anther thing I think we don't need to worry too much but should keep an eye for is that because we are waiting for 100 ms for receiving the packet for sniffing, the worst case time out is now increased to 500ms. It is still too early to say it would break anything, but I am ready to fix this will a separate timeout timer as needed.

I have following observations for your pcap:
* Domain sniffed from first 3 initial packets sent by client
* 3 initial packets sent within 1ms

The 100ms is controlled by `mb, _ := r.reader.ReadMultiBufferTimeout(time.Millisecond * 100)`. `ReadMultiBufferTimeout` return at least one packet in each read. In worst case, it may return only one packet in each call, making the 3rd initial packet not sniffed. 

So the decision is now based on whether `ReadMultiBufferTimeout` is guaranteed to return all subsequent initial packets in second call. Since initial packets are sent within 1ms, it seems very likely to happen.

Btw, I removed the increased attempt in my updated commit.

**Comment by Vigilans** [comment]
> (I'm trying to fix sniffing [fragmented TLS records](https://upb-syssec.github.io/blog/2023/record-fragmentation/) for the TLS sniffer, and found ReadMultiBufferTimeout only returns exactly one packet for TCP. See https://github.com/dyhkwong/v2ray-core/commit/3b7adc1fb98ee902bb5c2518bbdbd59d2245229d

I am organizing your commit into this PR. A very interesting observation occured: your update to `ReadClientHello` makes much of new tests in this PR failed:
```
--- FAIL: TestSniffQUICComplex (0.00s)
    --- FAIL: TestSniffQUICComplex/QUIC_Chromebook_Handshake[1];_packet_1 (0.00s)
        .../v2ray-core/common/protocol/quic/sniff_test.go:228: SniffQUIC() error = <nil>, wantErr true
    --- FAIL: TestSniffQUICComplex/QUIC_Chromebook_Handshake[2];_packet_1 (0.00s)
        .../v2ray-core/common/protocol/quic/sniff_test.go:228: SniffQUIC() error = <nil>, wantErr true
    --- FAIL: TestSniffQUICComplex/QUIC_Chromebook_Handshake[2];_packet_1-2 (0.00s)
        .../v2ray-core/common/protocol/quic/sniff_test.go:228: SniffQUIC() error = <nil>, wantErr true
    --- FAIL: TestSniffQUICComplex/QUIC_Chromebook_Handshake[3];_packet_1 (0.00s)
        .../v2ray-core/common/protocol/quic/sniff_test.go:228: SniffQUIC() error = <nil>, wantErr true
    --- FAIL: TestSniffQUICComplex/QUIC_Chromebook_Handshake[3];_packet_1-2 (0.00s)
        .../v2ray-core/common/protocol/quic/sniff_test.go:228: SniffQUIC() error = <nil>, wantErr true
    --- FAIL: TestSniffQUICComplex/QUIC_Chromebook_Handshake[4];_packet_1 (0.00s)
        .../v2ray-core/common/protocol/quic/sniff_test.go:228: SniffQUIC() error = <nil>, wantErr true
    --- FAIL: TestSniffQUICComplex/QUIC_Chromebook_Handshake[5];_packet_1 (0.00s)
        .../v2ray-core/common/protocol/quic/sniff_test.go:228: SniffQUIC() error = <nil>, wantErr true
    --- FAIL: TestSniffQUICComplex/QUIC_Chromebook_Handshake[7];_packet_1 (0.00s)
        .../v2ray-core/common/protocol/quic/sniff_test.go:228: SniffQUIC() error = <nil>, wantErr true
    --- FAIL: TestSniffQUICComplex/QUIC_Chromebook_Handshake[7];_packet_1-2 (0.00s)
        .../v2ray-core/common/protocol/quic/sniff_test.go:228: SniffQUIC() error = <nil>, wantErr true
```

This is because your update makes `ReadClientHello` read partial of QUIC's crypto data and discovered domain in it! Not all initial packets needs to be ready before domain can be sniffed out.

I think your update to TLS can be maintained in a separate PR, and let this PR focus on update of QUIC and your new `errRemainingData`?

**Comment by Vigilans** [comment]
I've updated with your commit. It looks generally solid, with some modifications and fixes:
* `errRemainingData` renamed to `ErrProtoNeedsMoreData`. It means "Protocol matches, but need more data to complete sniffing". In contrast, `ErrNoClue` means "Protocol not matches, and sniffer cannot determine whether there will be a match or not".
* `ErrProtoNeedsMoreData` also timeout with 2 failed attempts, but only error of `cReader.Cache` is counted as one failed attempt. This is just a more permissive version of your check. In contrast, each `ErrNoClude` counts as failed attempt.
*  Fixed `Sniffer.Sniff` logic in `app/dispatcher/sniffer.go`. Your commit did not handle `pendingSniffer`, which controls sniffers used in next attempt. Before this commit it only includes sniffers returning `ErrNoClue`. I've fixed it to exclusively use the sniffer if returning `ErrProtoNeedMoreData`, because this sniffer's protocol matched.

**Comment by Vigilans** [comment]
Btw, I noticed your comment in https://github.com/XTLS/Xray-core/issues/3363#issuecomment-2111868904, I've checked code, and it seems `sniffer.Sniff` is blocking `routedDispatch` call, so routing is happened only after all packets sniffing attempts are done. In this case the worse case 500ms timeout mentioned by @xiaokangwang seems more important now. 

In the meantime, it proves your update to TLS `ReadClientHello` worths another PR, because it effectively reduces required initial packets to sniff out domain (in optimal case only first packet is enough).

File: common/protocol/quic/sniff_test.go (hunk 2 above)
Comment (xiaokangwang) [reply common/protocol/quic/sniff_test.go:130]: Thanks for finding the cause... I did try to copy and paste multiple times but maybe I made the same mistake each time...

**Comment by xiaokangwang** [comment]
@Vigilans I think this pull request is ready to be merged. Do you think we should proceed with merge or there is any known issue you would like to fix before it get merged?

**Comment by Vigilans** [comment]
@xiaokangwang I've been running v2ray binary with new commits in this PR and did not witness issue after running for a day. I think we can proceed with the merge.

**Comment by xiaokangwang** [comment]
Thanks for your contribution. I will merge this now.

**Comment by Vigilans** [comment]
> [7f50592](https://github.com/v2fly/v2ray-core/commit/7f50592823d3c893c5bbb6e1292ef08ae76947be) caused high CPU comsumption and high battery usage. Reverting it can solve this. [pprof](https://github.com/user-attachments/files/18843984/profile001.pdf)

Confirmed my instance also got high CPU consumption. Looking into it.

**Comment by xiaokangwang** [comment]
See also: https://github.com/v2fly/v2ray-core/issues/3316
There might be a buffer corruption issue.

**Comment by xiaokangwang** [comment]
> See also: #3316 There might be a buffer corruption issue.
@Vigilans 
I saw your comment on the referenced issue. Maybe a full buffer analysis was necessary to find out the root cause of issue. Sadly my hope for quick victory didn’t last long.

**Comment by Vigilans** [comment]
> See also: #3316 There might be a buffer corruption issue.

Fix PR for it: https://github.com/v2fly/v2ray-core/pull/3320
